### PR TITLE
Plugin support

### LINKED ISSUE #353: Pluggable service running on the leader node
It would be useful if we can provide a pluggable service which is started when the node is elected as a leader and stopped when its role is changed to a follower. 
MirroringService is one of the candidates which can be implemented as a plugin.

### PR DESCRIPTION
**Pull request by hyangtack**
Motivation:
It would be useful if we can provide a plugin which is started when the node is elected as a leader and stopped when its role is changed to a follower.
MirroringService is one of the candidates which can be implemented as a plugin.

Modifications:
- Added `c.l.c.server.plugin` package.
  - Added `Plugin` interface to start/stop the plugin.
  - Added `PluginTarget` enum to specify the target replica that the plugin is executed on.
- Made `DefaultMirroringService` as a built-in plugin.
- Refactoring
  - Added `c.l.c.server.command`/`mirror`/`storage`/`storage.project`/`storage.repository` packages
    in order to make the following classes public:
    - `Command` and `CommandExecutor`
    - `Mirror`
    - `StorageManager`
    - `Project` and `ProjectManager`
    - `Repository` and `RepositoryManager`
- Misc
  - Configured Mockito to make a mock of `final` class/method.

Result:
- Closes #353

### CODE REVIEW

File: server/src/main/java/com/linecorp/centraldogma/server/storage/project/Project.java [hunk 1]
```diff
@@ -14,44 +14,61 @@
...
 
+/**
+ * A top-level element in Central Dogma storage model. A project has {@code dogma} and {@code meta}
+ * repositories by default which contain configurations accessible for administrators and project owners
+ * respectively.
+ */
 public interface Project {
     /**
-     * The repository that contains the internal settings of a project.
+     * The repository that contains configurations of a project, which are allowed to administrators.
```
Comment (trustin) [inline server/src/main/java/com/linecorp/centraldogma/server/storage/project/Project.java]: configurations of a project -> project configuration files ?

File: server/src/main/java/com/linecorp/centraldogma/server/storage/project/Project.java [hunk 2]
```diff
@@ -14,44 +14,61 @@
...
      */
     String REPO_DOGMA = "dogma";
 
     /**
-     * The repository that contains the user-modifiable settings of a project.
+     * The repository that contains configurations of a project, which are allowed to project owners.
```
Comment (trustin) [inline server/src/main/java/com/linecorp/centraldogma/server/storage/project/Project.java]: configurations of a project -> project configuration files ?

File: server/src/main/java/com/linecorp/centraldogma/server/storage/repository/FindOption.java [hunk 3]
```diff
@@ -14,12 +14,19 @@
...
+package com.linecorp.centraldogma.server.storage.repository;
 
 import java.util.Map;
 
+import javax.annotation.Nullable;
+
 import com.linecorp.centraldogma.internal.Util;
 
+/**
+ * An option which is used to find files.
```
Comment (trustin) [inline server/src/main/java/com/linecorp/centraldogma/server/storage/repository/FindOption.java]: which is specified when retrieving one or more files?

File: server/src/main/java/com/linecorp/centraldogma/server/storage/repository/MetaRepository.java [hunk 4]
```diff
@@ -14,12 +14,18 @@
...
-package com.linecorp.centraldogma.server.internal.storage.repository;
+package com.linecorp.centraldogma.server.storage.repository;
 
 import java.util.Set;
 
-import com.linecorp.centraldogma.server.internal.mirror.Mirror;
+import com.linecorp.centraldogma.server.mirror.Mirror;
 
+/**
+ * A Revision-controlled filesystem-like repository which is named {@code meta}.
```
Comment (trustin) [inline server/src/main/java/com/linecorp/centraldogma/server/storage/repository/MetaRepository.java]: `"meta"`?

File: server/src/test/java/com/linecorp/centraldogma/server/internal/plugin/PluginTest.java [hunk 5]
```diff
@@ -46,9 +46,9 @@
...
 import com.linecorp.centraldogma.common.Revision;
 import com.linecorp.centraldogma.common.ShuttingDownException;
+import com.linecorp.centraldogma.server.internal.storage.project.DefaultProject;
 import com.linecorp.centraldogma.server.internal.storage.project.DefaultProjectManager;
-import com.linecorp.centraldogma.server.internal.storage.project.Project;
-import com.linecorp.centraldogma.server.internal.storage.project.ProjectManager;
+import com.linecorp.centraldogma.server.storage.project.ProjectManager;
 
 @Ignore("Nashorn raises an AssertionError for an unknown reason.")
 public class PluginTest {
```
Comment (trustin) [inline server/src/test/java/com/linecorp/centraldogma/server/internal/plugin/PluginTest.java]: How about zapping this completely, including the JS-based plugin code?

File: server/src/main/java/com/linecorp/centraldogma/server/storage/StorageManager.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    /**
+     * Restores an element with the specified {@code name}.
+     *
+     * @param name the name of an element which is supposed to be restored
+     */
+    T unremove(String name);
+
+    /**
+     * Ensures that this manager is open.
+     */
```
Comment (minwoox) [inline server/src/main/java/com/linecorp/centraldogma/server/storage/StorageManager.java]: Perhaps it's better to specify that `IllegalStateException` and `CentralDogmaException` can be raised using `@throws`?

File: server/src/main/java/com/linecorp/centraldogma/server/PluginGroup.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            final List<CompletionStage<Void>> futures =
+                    plugins.stream().map(plugin -> plugin.start(arg)).collect(toImmutableList());
+            return CompletableFutures.allAsList(futures).thenApply(unused -> null);
+        }
+
+        @Override
+        protected CompletionStage<Void> doStop(@Nullable PluginContext arg) throws Exception {
+            assert arg != null;
+            final List<CompletionStage<Void>> futures =
+                    plugins.stream().map(plugin -> plugin.stop(arg)).collect(toImmutableList());
```
[the remark this answers is not in the document]
Comment (hyangtack) [reply server/src/main/java/com/linecorp/centraldogma/server/PluginGroup.java]: Added. PTAL :-)

File: server/src/main/java/com/linecorp/centraldogma/server/command/Command.java [hunk 8]
```diff
@@ -163,51 +308,93 @@
...
+     * Not implemented.
+     */
     static Command<Void> removePlugin(Author author, String projectName, String pluginName) {
         return removePlugin(null, author, projectName, pluginName);
     }
 
+    /**
+     * Not implemented.
+     */
     static Command<Void> removePlugin(@Nullable Long timestamp, Author author,
```
[the remark this answers is not in the document]
Comment (hyangtack) [reply server/src/main/java/com/linecorp/centraldogma/server/command/Command.java]: Removed! Thanks! 😄

File: server/src/test/java/com/linecorp/centraldogma/server/internal/plugin/PluginTest.java (hunk 5 above)
Comment (hyangtack) [reply server/src/test/java/com/linecorp/centraldogma/server/internal/plugin/PluginTest.java]: Removed! Thanks! 😃

File: server/src/main/java/com/linecorp/centraldogma/server/PluginGroup.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+     *
+     * @param classLoader which is used to load the {@link Plugin}s
+     * @param target the {@link PluginTarget} which would be loaded
+     */
+    @Nullable
+    static PluginGroup loadPlugins(ClassLoader classLoader, PluginTarget target,
+                                   CentralDogmaConfig config) {
+        requireNonNull(classLoader, "classLoader");
+        requireNonNull(target, "target");
+
```
[the remark this answers is not in the document]
Comment (hyangtack) [reply server/src/main/java/com/linecorp/centraldogma/server/PluginGroup.java]: Oops! Thanks!

File: server/src/main/java/com/linecorp/centraldogma/server/storage/StorageManager.java (hunk 6 above)
Comment (hyangtack) [reply server/src/main/java/com/linecorp/centraldogma/server/storage/StorageManager.java]: Added. PTAL :-)

File: server/build.gradle [hunk 10]
```diff
@@ -75,7 +75,7 @@ clientDependencies {
     }
 }
 
-task copyWebApp(type:Copy){
+task copyWebApp(type: Copy) {
```
Comment (trustin) [inline server/build.gradle]: Thanks :smile:

File: server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/DefaultMirroringServicePlugin.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        final DefaultMirroringService mirroringService = this.mirroringService;
+        if (mirroringService != null && mirroringService.isStarted()) {
+            mirroringService.stop();
+        }
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Override
+    public boolean isEnabled(CentralDogmaConfig config) {
+        return config.isMirroringEnabled();
```
Comment (minwoox) [inline server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/DefaultMirroringServicePlugin.java]: nit: `requireNonNull(config, "cofnig");`

File: server/src/main/java/com/linecorp/centraldogma/server/mirror/Mirror.java [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+     *                 If unspecified, {@code 0 * * * * ?} (every minute) is used.
+     * @param direction the direction of mirror
+     * @param credential the authentication credentials which are required when accessing the Git repositories
+     * @param localRepo the Central Dogma repository name
+     * @param localPath the directory path in the {@code localRepo}
+     * @param remoteUri the URI of the Git repository which will be mirrored from
+     */
+    static Mirror of(Cron schedule, MirrorDirection direction, MirrorCredential credential,
+                     Repository localRepo, String localPath, URI remoteUri) {
+        final String scheme = remoteUri.getScheme();
```
Comment (minwoox) [inline server/src/main/java/com/linecorp/centraldogma/server/mirror/Mirror.java]: nit: null check?

File: server/src/main/java/com/linecorp/centraldogma/server/mirror/MirrorUtil.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+public final class MirrorUtil {
+
+    static final Pattern DOGMA_PATH_PATTERN = Pattern.compile("^/([^/]+)/([^/]+)\\.dogma$");
+
+    /**
+     * Normalizes the specified {@code path}. A path which starts and ends with {@code /} would be returned.
+     * Also, it would not have consecutive {@code /}.
+     */
+    public static String normalizePath(String path) {
+        if (path.isEmpty()) {
```
Comment (minwoox) [inline server/src/main/java/com/linecorp/centraldogma/server/mirror/MirrorUtil.java]: nit: requreNonNull(path, "path");?

**Pull request merged**